library/asb_last_operation.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/python
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: asb_last_operation
+short_description: Adds a annotation to the pod running the apb with a last operation description
+description:
+     - Takes a string containting the last operation description this decription should be clear and concise. 
+       This description is then added an an annotation to the pod executing the apb and read by the broker.
+notes: []
+requirements: []
+author:
+    - "Red Hat, Inc."
+options:
+  description:
+    description:
+      - 'string describing the last operation'
+    required: true
+    default: ""
+env:
+        - Set via the downward API on the APB Pod
+        - name: ENV_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.name
+        - name: ENV_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace      
+'''
+
+EXAMPLES = '''
+- name: update last operation
+  asb_last_operation:
+    description:
+      "10%: creating service route"
+'''
+RETURN = '''
+
+'''
+
+import json
+import base64
+import os
+from ansible.module_utils.basic import AnsibleModule
+try:
+    from kubernetes import client, config
+    config.load_kube_config()
+    api = client.CoreV1Api()
+except Exception as error:
+    ansible_module.fail_json(msg="Error attempting to load kubernetes client: {}".format(error))
+
+LAST_OPERATION_ANNOTATION = "apb_last_operation"
+ENV_NAME = "POD_NAME"
+ENV_NAMESPACE = "POD_NAMESPACE"
+
+
+def main():
+
+    argument_spec = dict(
+        description=dict(required=True, type='str')
+    )
+
+    ansible_module = AnsibleModule(argument_spec=argument_spec)
+
+    if 'description' not in ansible_module.params:
+        error, 'last operation requires a description'
+    
+    if ENV_NAME not in os.environ:
+        error, 'expected a POD_NAME in the environment'
```

Comment:
What is actually happening here? This may be a case of ignorance, but if you could share output from a test-case without setting `POD_NAME` I would appreciate it.

Reply:
I am definitely not a python expert here, how would you expect to see the error handled? Also I think it is actually highly unlikely that these values would not be present as they are set by Ansible Broker. Either way I am happy to change it to meet your expectations

Reply:
:sunglasses: I'm not a python expert here either. You could follow https://github.com/ansibleplaybookbundle/ansible-asb-modules/blob/master/library/asb_encode_binding.py#L75-L79. That seems fair and predictable to me.

We are exploring the idea of running APBs outside of the Ansible Broker context, and some of the QE guys have already run into this. That is the reason for calling it out. You can run an apb that has the module with something like (you can just remove the environment vars to see if it would break).

```
docker run \
--rm \
--net=host \
-e POD_NAME=fake-pod-name \
-e POD_NAMESPACE=fake-pod-namespace \
-v $HOME/.kube:/opt/apb/.kube:z \
-u $UID \
docker.io/djzager/my-test-apb:latest \
provision \
--extra-vars 'namespace=fake-pod-namespace' \
--extra-vars '_apb_plan_id=dev'
```